x/poolmanager/client/query_proto_wrap.go

```diff
@@ -254,3 +258,378 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
+// For a balancer pool if an amount entered is greater than the total pool liquidity the trade estimated would be
+// the full liquidity of the other token. If the amount is small it would return a close 1:1 trade of the
+// smallest units.
+func (q Querier) estimateTradeBasedOnPriceImpactBalancerPool(
+	ctx sdk.Context,
+	req queryproto.EstimateTradeBasedOnPriceImpactRequest,
+	spotPrice, adjustedMaxPriceImpact sdk.Dec,
+	swapModule types.PoolModuleI,
+	poolI types.PoolI,
+) (*queryproto.EstimateTradeBasedOnPriceImpactResponse, error) {
+	// There isn't a case where the tokenOut could be zero or an error is received but those possibilities are handled
+	// anyway.
+	tokenOut, err := swapModule.CalcOutAmtGivenIn(ctx, poolI, req.FromCoin, req.ToCoinDenom, sdk.ZeroDec())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	if tokenOut.IsZero() {
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+			OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+		}, nil
+	}
+
+	// Validate if the trade as is respects the price impact, if it does re-estimate it with a swap fee and return
+	// the result.
+	currTradePrice := sdk.NewDec(req.FromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+	priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
+
+	if priceDeviation.LTE(adjustedMaxPriceImpact) {
+		tokenOut, err = swapModule.CalcOutAmtGivenIn(
+			ctx, poolI, req.FromCoin, req.ToCoinDenom, poolI.GetSpreadFactor(ctx),
+		)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  req.FromCoin,
+			OutputCoin: tokenOut,
+		}, nil
+	}
+
+	// Define low and high amount to search between. Start from 1 and req.FromCoin.Amount as initial range.
+	lowAmount := sdk.OneInt()
+	highAmount := req.FromCoin.Amount
+	currFromCoin := req.FromCoin
+
+	for lowAmount.LTE(highAmount) {
+		// Calculate currFromCoin as the new middle amount to try trade.
+		midAmount := lowAmount.Add(highAmount).Quo(sdk.NewInt(2))
+		currFromCoin = sdk.NewCoin(req.FromCoin.Denom, midAmount)
+
+		// There isn't a case where the tokenOut could be zero or an error is received but those possibilities are
+		// handled anyway.
+		tokenOut, err := swapModule.CalcOutAmtGivenIn(
+			ctx, poolI, currFromCoin, req.ToCoinDenom, sdk.ZeroDec(),
+		)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		if tokenOut.IsZero() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+			}, nil
+		}
+
+		currTradePrice := sdk.NewDec(currFromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+		priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
```

Comment:
I think we can make a separate method for this since this is re-occurring & for code readability

Reply:
Added